paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/sources/PaimonMicroBatchStream.scala

```diff
@@ -0,0 +1,182 @@
+import org.apache.paimon.table.source.ReadBuilder
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.connector.read.{InputPartition, PartitionReaderFactory}
+import org.apache.spark.sql.connector.read.streaming.{MicroBatchStream, Offset}
+
+import scala.util.Try
+
+class PaimonMicroBatchStream(
+    originTable: FileStoreTable,
+    readBuilder: ReadBuilder,
+    checkpointLocation: String)
+  extends MicroBatchStream
+  with StreamHelper
+  with Logging {
+
+  private var _table: FileStoreTable = originTable
+
+  private lazy val options = Options.fromMap(table.options())
+
+  private lazy val currentSnapshot = Option(table.snapshotManager().latestSnapshot())
+
+  private val (startSnapshotId, withStartSnapshot) = checkAndInitScanOptions()
+
+  private var committedOffset: Option[PaimonSourceOffset] = None
+
+  lazy val initOffset: PaimonSourceOffset = {
+    PaimonSourceOffset(startSnapshotId, -1, withStartSnapshot)
+  }
+
+  override def latestOffset(): Offset = {
+    getLatestOffset
+  }
+
+  override def planInputPartitions(start: Offset, end: Offset): Array[InputPartition] = {
+    val startOffset = {
+      val startOffset0 = PaimonSourceOffset.apply(start)
+      if (startOffset0.compareTo(initOffset) < 0) {
+        initOffset
+      } else {
+        startOffset0
+      }
+    }
+    val endOffset = PaimonSourceOffset.apply(end)
+
+    getBatch(startOffset, endOffset)
+      .map(ids => new SparkInputPartition(ids.entry))
+      .toArray[InputPartition]
+  }
+
+  override def createReaderFactory(): PartitionReaderFactory = {
+    new SparkReaderFactory(readBuilder)
+  }
+
+  override def initialOffset(): Offset = {
+    initOffset
+  }
+
+  override def deserializeOffset(json: String): Offset = {
+    PaimonSourceOffset.apply(json)
+  }
+
+  override def commit(end: Offset): Unit = {
+    committedOffset = Some(PaimonSourceOffset.apply(end))
+    logInfo(s"$committedOffset is committed.")
+  }
+
+  override def stop(): Unit = {}
+
+  override def table: FileStoreTable = _table
+
+  /** Parse [[Options]] to initialize the state that used to generate the init offset. */
+  private def checkAndInitScanOptions(): (Long, Boolean) = {
+    val scanSnapshotIdOptional = options.getOptional(SCAN_SNAPSHOT_ID)
+    val scanTSOptional = options.getOptional(SCAN_TIMESTAMP_MILLIS)
+    if (scanSnapshotIdOptional.isPresent && scanTSOptional.isPresent) {
+      throw new IllegalArgumentException(
+        "scan.snapshot-id and scan.timestamp-millis can be provided at the same time.")
+    }
+
+    // choose the bigger one between the earliest snapshotId and the provided snapshotId if exists as the init snapshotId
+    val earliestSnapshotId = Try(table.snapshotManager().earliestSnapshotId())
+      .map(_.longValue())
+      .getOrElse(Long.MinValue)
+
+    var scanMode = FROM_SNAPSHOT
+    var startSnapshotId = currentSnapshot.map(_.id()).getOrElse(1L)
+    var withStartSnapshot = false
+    options.get(SCAN_MODE) match {
```

Comment:
Can we ignore `SCAN_MODE` here?
We may support more scan modes in the future, but if we decouple here, it will be difficult to maintain.

When initializing for the first time, can we have a `null` offset so that we don't have to forcibly generate a startOffset.

Reply:
Sorry that I can't understand why we should have a null offset instead of an initialized offset.
Spark will call  `initialOffset` to get the starting offset which is parsed by `SCAN_MODE` and other related options.

Reply:
And i don't think we can support other scan modes without changing any codes.

Reply:
Logically speaking, Spark does not need to perceive specific scan modes, for example, it only generates offsets at the beginning: `(startOffset: Nil, endOffset: $last_snapshot)`.
Then, during scan, if startOffset is Nil, directly call StreamTableScan.scan.
At this point, obtain the snapshotId to which the plan belongs, set offset to: snapshotId + splitNumber + isInitial: true.

This process does not require Spark to perceive scan modes.

Reply:
Understood. I still think we need to initialize `initialOffset` by parsing `ScanMode` and other related configs. Cause:
1. `initialOffset` will return an accurate offset, not just `null`, due to the definition of the method.
2. As `startOffset` in the first batch, `initialOffset` will also be used in many calls, like `Offset latestOffset(Offset startOffset, ReadLimit limit)` after supporting read limit.
3. Cause more complicated logic, just to avoid perceive scan mode. maybe some unexpected errors will occur.

Reply:
I still adhere to my viewpoint for the following reasons:

I think it just depends on our definition, and the initialization state is also an initialOffset. I don't necessarily mean to use null to represent it, it depends on the abstract space provided by the Spark interface.

When using StreamTableScan, we don't have any requirements to couple scan modes in Spark code, which is strange and will make it increasingly troublesome later on.
When we added a scan mode (which is common and has occurred multiple times in the past), we found that we need to modify Spark's code, while Flink's code does not need to be modified, which is strange.

Reply:
Your point is avoiding to perceive scan mode just for spark interface, while my point is that `initialOffset` should be correct and meaningful. I have an idea to meet these: let `StreamTableScan` return this starting snapshot id and whether it'll scan the full snapshot.

Reply:
We can define initialOffset is something like `-unbounded`, because Paimon store all data for snapshot scan. Is this possible?

Reply:
related to https://github.com/apache/incubator-paimon/pull/1955.